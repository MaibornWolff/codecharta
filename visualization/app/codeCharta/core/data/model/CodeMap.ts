export interface CodeMapNode {
    name: string,
    type: string,
    children?: CodeMapNode[]
    attributes: {
        [key: string]: number
    };
    deltas?: {
        [key: string]: number
    };
    link?: string;
    origin?: string;
    visible?: boolean;
    path?: string;
    markingColor?: string;
}

export interface CodeMap {

    fileName: string,
    apiVersion?: string,
    projectName: string,
    root: CodeMapNode,
    dependencies?: {
        static?: CodeMapDependency[],
        temporal_coupling?: CodeMapDependency[]
    }

}

export interface CodeMapDependency {
    node: string,
    nodeFilename?: string,
    dependantNode: string,
    dependantNodeFilename?: string,
    pairingRate?: number,
    averageRevs?: number
    visible: boolean
}