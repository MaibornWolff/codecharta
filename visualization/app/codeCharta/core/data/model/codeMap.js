export class CodeMap {

    constructor(fileName, projectName, root) {

        this.fileName = fileName;
        this.projectName = projectName;
        this.root = root;

    }


}