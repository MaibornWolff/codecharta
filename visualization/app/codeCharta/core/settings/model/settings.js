/**
 * This class defines the available set of settings in the CodeCharta application
 */
export class Settings {

    /**
     * @constructor
     * @param {CodeMap} map
     * @param {Range} neutralColorRange
     * @param {string} areaMetric
     * @param {string} heightMetric
     * @param {string} colorMetric
     * @param {boolean} deltas
     * @param {Number} amountOfTopLabels
     * @param {Scale} scaling
     * @param {Scale} camera
     * @param {Number} margin
     */
    constructor(map, neutralColorRange, areaMetric, heightMetric, colorMetric, deltas, amountOfTopLabels, scaling, camera, margin) {

        /**
         * currently selected map
         * @type {CodeMap}
         */
        this.map = map;

        /**
         * currently set neutral color range. This defines the yellow color range and the whether higher or lower values are green or red.
         * @type {Range}
         */
        this.neutralColorRange = neutralColorRange;

        /**
         * currently selected area metric
         * @type {string}
         */
        this.areaMetric = areaMetric;

        /**
         * currently selected heightm metric
         * @type {string}
         */
        this.heightMetric = heightMetric;

        /**
         * currently selected color metric
         * @type {string}
         */
        this.colorMetric = colorMetric;

        /**
         * delta flag
         * @type {boolean}
         */
        this.deltas = deltas;

        /**
         * number of highest buildings with labels
         * @type {Number}
         */
        this.amountOfTopLabels = amountOfTopLabels;

        /**
         * scaling settings
         * @type {Scale}
         */
        this.scaling = scaling;

        /**
         * camera position
         * @type {Scale}
         */
        this.camera = camera;

        /**
         * margin between buildings
         * @type {number}
         */
        this.margin = margin;

    }

    /**
     * Imports the given settings values without replacing this object with the given one.
     * @param {Settings} settings given settings
     */
    importSettingValues(settings){
        this.map = settings.map;
        this.neutralColorRange = settings.neutralColorRange;
        this.areaMetric = settings.areaMetric;
        this.heightMetric = settings.heightMetric;
        this.colorMetric = settings.colorMetric;
        this.deltas = settings.deltas;
        this.amountOfTopLabels = settings.amountOfTopLabels;
        this.scaling = settings.scaling;
        this.camera = settings.camera;
        this.margin = settings.margin;
    }

}
