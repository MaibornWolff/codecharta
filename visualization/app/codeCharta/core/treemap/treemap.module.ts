"use strict"

import angular from "angular"

import "../data/data.module"
import { TreeMapService } from "./treemap.service"

angular.module("app.codeCharta.core.treemap", ["app.codeCharta.core.data"])

angular.module("app.codeCharta.core.treemap").service("treeMapService", TreeMapService)
