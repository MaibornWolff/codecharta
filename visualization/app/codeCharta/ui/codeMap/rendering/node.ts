export interface NodeAttributes {
    readonly [key : string] : number;
}

export interface Node {
    name : string;
    width : number;
    height : number;
    length : number;
    depth : number;
    x0 : number;
    z0 : number;
    y0 : number;
    isLeaf : boolean;
    deltas : NodeAttributes;
    attributes : NodeAttributes;
    children : Node[];
    parent : Node;
    heightDelta : number;
    visible : boolean;
    path : string;
    origin : string;
    link : string;
    markingColor: string;
    flat: boolean;
}