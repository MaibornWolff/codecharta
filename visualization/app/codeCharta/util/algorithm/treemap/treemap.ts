import BoundingBox from "../streetLayout/boundingBox"
import { CodeMapNode } from "../../../codeCharta.model"
import { StreetLayoutValuedCodeMapNode } from "../../streetLayoutGenerator"
import Point from "../point"
import { StreetLayoutHelper } from "../../streetLayoutHelper"

export default abstract class Treemap extends BoundingBox {
	protected treeMapNodes: StreetLayoutValuedCodeMapNode[] = []
	protected metricName: string

	constructor(rootNode: CodeMapNode) {
		super(rootNode)
	}

	public abstract layout(origin: Point, margin: number): StreetLayoutValuedCodeMapNode[]

	public calculateDimension(metricName: string): void {
		this.metricName = metricName
		this.metricValue = StreetLayoutHelper.calculateSize(this.node, metricName)
		this.width = Math.sqrt(this.metricValue)
		this.height = Math.sqrt(this.metricValue)
	}
}
