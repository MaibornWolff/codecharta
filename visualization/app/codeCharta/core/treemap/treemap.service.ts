import {CodeMapNode, CodeMapDependency} from "../data/model/CodeMap";
import {node} from "../../ui/codeMap/rendering/node";
import {DataService} from "../data/data.service";
import * as d3 from "d3";
import {TreeMapUtils} from "./treemap.util";
import {HierarchyNode} from "d3";

export interface ValuedCodeMapNode {
    data: CodeMapNode;
    children?: ValuedCodeMapNode[];
    value: number;
}

export interface SquarifiedValuedCodeMapNode {
    data: CodeMapNode;
    children?: SquarifiedValuedCodeMapNode[];
    parent?: SquarifiedValuedCodeMapNode;
    value: number;
    x0: number;
    y0: number;
    x1: number;
    y1: number;
}

export interface TreeMapSettings {
    size: number;
    areaKey: string;
    heightKey: string;
    margin: number;
    invertHeight: boolean;
    visibleTemporalCouplingDependencies: CodeMapDependency[];
    useCouplingHeight: boolean;
}

export class TreeMapService {

    private static HEIGHT_DIVISOR = 1;
    private static FOLDER_HEIGHT = 2;
    private static MIN_BUILDING_HEIGHT = 2;
    private static HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND = 0;
    private static PADDING_SCALING_FACTOR = 0.4;

    /* @ngInject */
    constructor(private dataService: DataService) {}

    public createTreemapNodes(data: CodeMapNode, s: TreeMapSettings): node {
        const squarified: SquarifiedValuedCodeMapNode = this.squarify(data, s);
        const heighted = this.addMapScaledHeightDimensionAndFinalizeFromRoot(squarified, s);
        return heighted;
    }

    private squarify(data: CodeMapNode, s: TreeMapSettings): SquarifiedValuedCodeMapNode {
        let root: HierarchyNode<CodeMapNode> = d3.hierarchy<CodeMapNode>(data);
        let nodesPerSide = 2 * Math.sqrt(root.descendants().length);
        let treeMap = d3.treemap<CodeMapNode>()
            .size([s.size + nodesPerSide*s.margin, s.size + nodesPerSide*s.margin])
            .paddingOuter(s.margin * TreeMapService.PADDING_SCALING_FACTOR || 1)
            .paddingInner(s.margin * TreeMapService.PADDING_SCALING_FACTOR || 1);

        return treeMap(root.sum((node) => this.calculateValue(node, s.areaKey))) as SquarifiedValuedCodeMapNode;
    }

    private addMapScaledHeightDimensionAndFinalizeFromRoot(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings): node {
        const heightScale = s.size / TreeMapService.HEIGHT_DIVISOR / this.dataService.getMaxMetricInAllRevisions(s.heightKey);

        const maxHeight = this.dataService.getMaxMetricInAllRevisions(s.heightKey);
        return this.addHeightDimensionAndFinalize(squaredNode, s, heightScale, maxHeight);
    }

    private addHeightDimensionAndFinalize(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings, heightScale: number, maxHeight: number, depth = 0, parent: node = null): node {

        let attr = squaredNode.data.attributes || {};
        let heightValue = attr[s.heightKey];

        if (heightValue === undefined || heightValue === null) {
            heightValue = TreeMapService.HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND;
        }

        const finalNode = TreeMapUtils.buildNodeFrom(squaredNode, heightScale, heightValue, maxHeight, depth, parent, s, TreeMapService.MIN_BUILDING_HEIGHT, TreeMapService.FOLDER_HEIGHT);

        if (squaredNode.children && squaredNode.children.length > 0) {
            const finalChildren: node[] = [];
            for (let i = 0; i < squaredNode.children.length; i++) {
                finalChildren.push(this.addHeightDimensionAndFinalize(squaredNode.children[i], s, heightScale, maxHeight, depth + 1, finalNode));
            }
            finalNode.children = finalChildren;
        }
        return finalNode;

    }

    private calculateValue(node: CodeMapNode, key: string): number {
        let result = 0;
        if (node.attributes && (!node.children || node.children.length === 0)) {
            result = node.attributes[key] || 0;
        }
        return result;
    }

}

