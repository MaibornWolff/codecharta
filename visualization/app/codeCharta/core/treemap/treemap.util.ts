import {SquarifiedValuedCodeMapNode, TreeMapSettings} from "./treemap.service";
import {node} from "../../ui/codeMap/rendering/node";

export class TreeMapUtils {

    public static countNodes(node: { children?: any }): number {
        let count = 1;
        if (node.children && node.children.length > 0) {
            for (let i = 0; i < node.children.length; i++) {
                count += TreeMapUtils.countNodes(node.children[i]);
            }
        }
        return count;
    }

    public static isNodeLeaf(node: { children?: any }): boolean {
        return !(node.children && node.children.length > 0);
    }

    public static buildNodeFrom(squaredNode: SquarifiedValuedCodeMapNode,
                                heightScale: number,
                                heightValue: number,
                                maxHeight: number,
                                depth: number,
                                parent: node,
                                s: TreeMapSettings,
                                minHeight: number,
                                folderHeight: number): node {

        if (s.invertHeight) {
            heightValue = (maxHeight - heightValue);
        }

        if (s.visibleTemporalCouplingDependencies && s.visibleTemporalCouplingDependencies.length > 0) {
            heightValue = this.getTemporalCouplingHeight(squaredNode, s, maxHeight, heightValue);
        }

        return {
            name: squaredNode.data.name,
            width: squaredNode.x1 - squaredNode.x0,
            height: Math.abs(TreeMapUtils.isNodeLeaf(squaredNode) ? Math.max(heightScale * heightValue, minHeight) : folderHeight),
            length: squaredNode.y1 - squaredNode.y0,
            depth: depth,
            x0: squaredNode.x0,
            z0: depth * folderHeight,
            y0: squaredNode.y0,
            isLeaf: TreeMapUtils.isNodeLeaf(squaredNode),
            attributes: squaredNode.data.attributes,
            deltas: squaredNode.data.deltas,
            parent: parent,
            heightDelta: Math.abs(squaredNode.data.deltas && squaredNode.data.deltas[s.heightKey] ? heightScale * squaredNode.data.deltas[s.heightKey] : 0),
            visible: squaredNode.data.visible,
            path: squaredNode.data.path,
            origin: squaredNode.data.origin,
            link: squaredNode.data.link,
            children: [],
            markingColor: parseInt(squaredNode.data.markingColor)
        };

    }

    private static getTemporalCouplingHeight(squaredNode: SquarifiedValuedCodeMapNode, s: TreeMapSettings, maxHeight: number, heightValue: number) {

        const NON_TEMPORAL_COUPLING_HEIGHT = 0;

        for (var couple of s.visibleTemporalCouplingDependencies) {

            if (squaredNode.data.path === couple.node ||
                squaredNode.data.path === couple.dependantNode) {
                return (s.useCouplingHeight) ? maxHeight / 100 * couple.pairingRate : heightValue;
            }
        }
        return NON_TEMPORAL_COUPLING_HEIGHT;
    }
}