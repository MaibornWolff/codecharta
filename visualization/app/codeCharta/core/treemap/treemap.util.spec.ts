import {TreeMapUtils} from "./treemap.util";
import {SquarifiedValuedCodeMapNode, TreeMapSettings} from "./treemap.service";
import {CodeMapDependency, CodeMapNode} from "../data/model/CodeMap";
import {node} from "../../ui/codeMap/rendering/node";
describe("treemap utils", () => {

    describe("build node", () => {

        let codeMapNode: CodeMapNode;
        let squaredNode: SquarifiedValuedCodeMapNode;
        let treeMapSettings: TreeMapSettings;

        let heightScale = 1;
        let heightValue = 100;
        let depth = 0;
        let parent = null;
        let minHeight = 1;
        let folderHeight = 2;
        let maxHeight = 2000;

        beforeEach(() => {

            codeMapNode = {
                name: "Anode",
                path: "/root/Anode",
                type: "File",
                attributes: {}
            }

            squaredNode = {
                data: codeMapNode,
                value: 42,
                x0: 0,
                y0: 0,
                x1: 400,
                y1: 400
            }

            treeMapSettings = {
                size: 1,
                areaKey: "theArea",
                heightKey: "theHeight",
                margin: 15,
                invertHeight: false,
                visibleTemporalCouplingDependencies: [],
                useCouplingHeight: true
            }

        });

        function buildNode() {
            return TreeMapUtils.buildNodeFrom(
                squaredNode,
                heightScale,
                heightValue,
                maxHeight,
                depth,
                parent,
                treeMapSettings,
                minHeight,
                folderHeight
            );
        }

        it("minimal", () => {
            expect(buildNode()).toMatchSnapshot();
        });

        it("invertHeight", () => {
            treeMapSettings.invertHeight = true;
            expect(buildNode()).toMatchSnapshot();
        });

        it("deltas", () => {
            squaredNode.data.deltas = {};
            squaredNode.data.deltas[treeMapSettings.heightKey] = 33;
            expect(buildNode()).toMatchSnapshot();
            squaredNode.data.deltas = undefined;
        });

        it("is leaf", () => {
            let tmp = TreeMapUtils.isNodeLeaf;
            TreeMapUtils.isNodeLeaf = jest.fn();
            TreeMapUtils.isNodeLeaf.mockReturnValue(true);
            expect(buildNode()).toMatchSnapshot();
            TreeMapUtils.isNodeLeaf = tmp;
        });

        it("dependency couple node should use pairingrate/normal metric as height", () => {
            treeMapSettings.visibleTemporalCouplingDependencies = [{
                node: "/root/Anode",
                nodeFilename: "Anode",
                dependantNode: "/root/AnotherNode",
                dependantNodeFilename: "AnotherNode",
                pairingRate: 68,
                averageRevs: 17,
                visible: true,
            }];
            treeMapSettings.useCouplingHeight = true;
            expect(buildNode()).toMatchSnapshot();
            treeMapSettings.useCouplingHeight = false;
            expect(buildNode()).toMatchSnapshot();
        });

        it("should set lowest possible height caused by other visible dependency pairs", () => {
            treeMapSettings.visibleTemporalCouplingDependencies = [{
                node: "/root/AnotherNode1",
                nodeFilename: "AnotherNode1",
                dependantNode: "/root/AnotherNode2",
                dependantNodeFilename: "AnotherNode2",
                pairingRate: 33,
                averageRevs: 12,
                visible: true,
            }];
            expect(buildNode()).toMatchSnapshot();
        });

    });

    describe("detect leaves", () => {

        it("root with no children is leaf", () => {
            const root = {};
            expect(TreeMapUtils.isNodeLeaf(root)).toBeTruthy();
        });

        it("root plus child should not be a leaf", () => {
            const root = {children: [{}]};
            expect(TreeMapUtils.isNodeLeaf(root)).toBeFalsy();
            expect(TreeMapUtils.isNodeLeaf(root.children[0])).toBeTruthy();
        });

        it("root plus child in child", () => {
            const root = {children: [{children: [{}]}]};
            expect(TreeMapUtils.isNodeLeaf(root)).toBeFalsy();
            expect(TreeMapUtils.isNodeLeaf(root.children[0])).toBeFalsy();
            expect(TreeMapUtils.isNodeLeaf(root.children[0].children[0])).toBeTruthy();
        });

        it("root plus two children", () => {
            const root = {children: [{}, {}]};
            expect(TreeMapUtils.isNodeLeaf(root)).toBeFalsy();
            expect(TreeMapUtils.isNodeLeaf(root.children[0])).toBeTruthy();
            expect(TreeMapUtils.isNodeLeaf(root.children[1])).toBeTruthy();
        });

    });

    describe("count nodes", () => {

        it("root only should be 1", () => {
            const root = {};
            expect(TreeMapUtils.countNodes(root)).toBe(1);
        });

        it("root plus child should be 2", () => {
            const root = {children: [{}]};
            expect(TreeMapUtils.countNodes(root)).toBe(2);
        });

        it("root plus child in child should be 3", () => {
            const root = {children: [{children: [{}]}]};
            expect(TreeMapUtils.countNodes(root)).toBe(3);
        });

        it("root plus two children should be 3", () => {
            const root = {children: [{}, {}]};
            expect(TreeMapUtils.countNodes(root)).toBe(3);
        });

    });

}