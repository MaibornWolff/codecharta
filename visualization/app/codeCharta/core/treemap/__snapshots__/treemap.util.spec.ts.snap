// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treemap utils build node deltas 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": Object {
    "theHeight": 33,
  },
  "depth": 0,
  "height": 100,
  "heightDelta": 33,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": NaN,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap utils build node dependency couple node should use pairingrate/normal metric as height 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "height": 1360,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": NaN,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap utils build node dependency couple node should use pairingrate/normal metric as height 2`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "height": 100,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": NaN,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap utils build node invertHeight 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "height": 1900,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": NaN,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap utils build node is leaf 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "height": 100,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": NaN,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap utils build node minimal 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "height": 100,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": NaN,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treemap utils build node should set lowest possible height caused by other visible dependency pairs 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "height": 1,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": NaN,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;
